### PR TITLE
[Infra] Only compile core extension headers once

### PR DESCRIPTION
**Pull request by ncooke3**
A custom module map was too difficult. I couldn't get CocoaPods to respect the custom module map and stopped when I started to spin on things. I switched to guarding the headers against being included twice. It seems promising so far. I also tried with using `#pragma once` but that didn't work because (my guess) is that the problem is _not_ that the _same_ header is being included twice, it's that two different headers with the same content are _both_ being included (once in FirebaseCore and once in FirebaseCoreExtension).

Test cases and their results below:

1.  Below Podfile uses the current state and fails when building docs.
```
platform :ios, '13.0'

source 'https://github.com/firebase/SpecsDev.git'
source 'https://github.com/firebase/SpecsStaging.git'
source 'https://cdn.cocoapods.org/'

target 'tempDoc' do
  use_frameworks!

  # Pods for tempDoc
  pod 'FirebaseCrashlytics', '11.5.0'
end
```

2. Below Podfile uses updated `FirebaseCore` spec but `FirebaseCoreExtension` @ main. The framework target builds but the docs target does not build because of the `FirebaseCoreExtension` @ main does not have guarded headers.
```
platform :ios, '13.0'

source 'https://github.com/firebase/SpecsDev.git'
source 'https://github.com/firebase/SpecsStaging.git'
source 'https://cdn.cocoapods.org/'

target 'tempDoc' do
  use_frameworks!

  # Pods for tempDoc
  pod 'FirebaseCore', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCrashlytics', '11.5.0'
end
```

3. Below Podfile uses updated `FirebaseCore` spec AND updated `FirebaseCoreExtension` spec. The framework target builds AND the docs target builds.
```
platform :ios, '13.0'

source 'https://github.com/firebase/SpecsDev.git'
source 'https://github.com/firebase/SpecsStaging.git'
source 'https://cdn.cocoapods.org/'

target 'tempDoc' do
  use_frameworks!

  # Pods for tempDoc
  pod 'FirebaseCore', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCoreExtension', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCrashlytics', '11.5.0'
end
```

4. Below Podfile uses **old** product pod and floats to pick up updated `FirebaseCore` (the floating case). The framework target builds but the docs target does not build because of using the `FirebaseCoreExtension` @ main.
```
platform :ios, '13.0'

source 'https://github.com/firebase/SpecsDev.git'
source 'https://github.com/firebase/SpecsStaging.git'
source 'https://cdn.cocoapods.org/'

target 'tempDoc' do
  use_frameworks!

  # Pods for tempDoc
  pod 'FirebaseCore', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCrashlytics', '11.4.0' # older pod!
end
```

5. Below Podfile uses **old** product pod and floats to pick up updated `FirebaseCore` (the floating case). In this example, it floats to pick up FirebaseCoreExtension as well. The framework target builds and the docs target builds. 
```
platform :ios, '13.0'

source 'https://github.com/firebase/SpecsDev.git'
source 'https://github.com/firebase/SpecsStaging.git'
source 'https://cdn.cocoapods.org/'

target 'tempDoc' do
  use_frameworks!

  # Pods for tempDoc
  pod 'FirebaseCore', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCoreExtension', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls' 
  pod 'FirebaseCrashlytics', '11.4.0' # older pod!
end
```

### CODE REVIEW

**Comment by paulb777** [comment]
Are the podspecs published at `/Users/nickcooke/Developer/firebase-3` so they show up in the Pods folder instead of the Development Pods folder?

**Comment by ncooke3** [comment]
> Are the podspecs published at `/Users/nickcooke/Developer/firebase-3` so they show up in the Pods folder instead of the Development Pods folder?

No, is there a way I can do that locally? Or should I stage these to SpecsDev?

**Comment by paulb777** [comment]
I'm not aware of how to do it locally. Staging to SpecsDev would be a good way to test

**Comment by ncooke3** [comment]
Hmm, I actually don't think SpecsDev is that helpful because the spec points the same source that staging does.

**Comment by paulb777** [comment]
Should be fine to move the tags to the tip of your branch temporarily and use SpecsStaging

**Comment by paulb777** [comment]
Or  CocoaPods-11.5.0 should be sufficient

**Comment by ncooke3** [comment]
I actually went with this:
```ruby
 pod 'FirebaseCore', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls'
 pod 'FirebaseCoreExtension', :git => 'https://github.com/firebase/firebase-ios-sdk.git', :branch => 'nc/dup-decls'
 pod 'FirebaseCrashlytics', '11.4.0'
```

Screenshot 2024-10-29 at 6 35 19 PM

File: Gemfile [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # To update, change version below, run bundle install, test,
+# touch
```
Comment (paulb777) [inline Gemfile:2]: delete

**Review by paulb777 — APPROVED** [review]
LGTM

**Comment by ncooke3** [comment]
Merged and moving tags to get coverage in the nightlies.

**Comment by paulb777** [comment]
Confirmed the repro in  #13756 now successfully builds docs when pointing at SpecsStaging.

**Comment by andrewheard** [comment]
Nice solution, @ncooke3!

**Pull request merged**